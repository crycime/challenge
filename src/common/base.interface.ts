import { ModelFindOneAndUpdateOptions, ModelPopulateOptions } from 'mongoose';

export interface ModelFindInputInterface {
  filter: any;
  projection?: any;
  options?: {
    sort?: any;
    limit?: number;
    skip?: number;
    populates?: ModelPopulateOptions[];
    [key: string]: any;
  };
}

export interface ModelFindOneInputInterface {
  filter: any;
  projection?: any;
  populates?: ModelPopulateOptions[];
  options?: {
    [key: string]: any;
  };
}

export interface ModelAggregateInputInterface {
  pipeline: any[];
}

export interface ModelCountInputInterface {
  filter: any;
}

export interface ModelCreateInputInterface<T> {
  docs: Partial<T> | Partial<T>[];
  populates?: ModelPopulateOptions[];
}

export interface UpdateInputInterface<T> {
  conditions: any;
  update: Partial<T> & {
    $inc?: { [key: string]: number };
    $unset?: { [key: string]: number };
  };
  options?: ModelFindOneAndUpdateOptions;
  populates?: ModelPopulateOptions[];
}

export interface DeleteInputInterface {
  conditions: any;
  options?: ModelFindOneAndUpdateOptions;
  populates?: ModelPopulateOptions[];
}
