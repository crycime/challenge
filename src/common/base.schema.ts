import { ApiProperty } from '@nestjs/swagger';
import { IsDate, IsString } from 'class-validator';
import { Prop, Schema } from '@nestjs/mongoose';
import { Exclude, Expose } from 'class-transformer';
import { ConvertDate } from '../decorators/dto/convertDate.dto.decorator';

@Exclude()
@Schema()
export class BasicFields {
  @Expose()
  @ApiProperty()
  @ConvertDate('createdTime')
  @IsDate()
  createdTime: number;

  @Expose()
  @ApiProperty()
  @ConvertDate('updatedTime')
  @IsDate()
  updatedTime: number;

  @Expose()
  @ApiProperty()
  @IsString()
  @Prop({ type: String, default: 'system', required: true })
  createdBy: string;

  @Expose()
  @ApiProperty()
  @IsString()
  @Prop({ type: String, default: 'system', required: true })
  updatedBy: string;
}
