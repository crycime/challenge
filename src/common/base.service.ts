import * as _ from 'lodash';
import { Document, Model } from 'mongoose';
import {
  DeleteInputInterface,
  ModelAggregateInputInterface,
  ModelCountInputInterface,
  ModelCreateInputInterface,
  ModelFindInputInterface,
  ModelFindOneInputInterface,
  UpdateInputInterface,
} from './base.interface';
import { CatchError } from '../decorators/catchError.decorator';
import { MongoResponseErrorException } from '../exceptions/mongoResponseError.exception';
import { ConvertMongoObjectIds } from '../decorators/convertMongoObjectIds.decorator';
import { populateHelper } from '../utils/model.util';

@CatchError(MongoResponseErrorException)
@ConvertMongoObjectIds()
export abstract class ModelBaseService<T extends Document> {
  protected constructor(private readonly _model: Model<T>) {}

  find<K>({
    filter,
    projection,
    options,
  }: ModelFindInputInterface): Promise<T[]> {
    const docsQuery = this._model.find(
      filter,
      projection,
      _.omit(options, ['populates']),
    );
    return populateHelper<T>(docsQuery, _.get(options, 'populates')).lean();
  }

  findOne<K>({
    filter,
    projection,
    populates,
    options,
  }: ModelFindOneInputInterface): Promise<T> {
    const docsQuery = this._model.findOne(filter, projection, options);
    return populateHelper<T>(docsQuery, populates).lean();
  }

  aggregate<K>({ pipeline }: ModelAggregateInputInterface): Promise<K[]> {
    return this._model.aggregate(pipeline);
  }

  count({ filter }: ModelCountInputInterface): Promise<number> {
    return this._model.countDocuments(filter);
  }

  async create({ docs, populates }: ModelCreateInputInterface<T>): Promise<T> {
    const createdDoc = await this._model.create(docs);
    return populateHelper(createdDoc, populates).execPopulate();
  }

  async bulkCreate({ docs }: { docs: Partial<T>[] }): Promise<T[]> {
    return await this._model.create(docs);
  }

  async bulkWrite(docs: any): Promise<any> {
    return await this._model.bulkWrite(docs);
  }

  update({
    conditions,
    update,
    options,
    populates,
  }: UpdateInputInterface<T>): Promise<T> {
    const updatedDoc = this._model.findOneAndUpdate(
      conditions,
      {
        ...update,
        $inc: { ..._.get(update, '$inc', {}), __v: 1 },
      },
      {
        ...options,
        new: true,
      },
    );
    return populateHelper<T>(updatedDoc, populates).lean();
  }

  updateMany({
    conditions,
    update,
    options,
  }: UpdateInputInterface<T>): Promise<T> {
    return this._model.updateMany(
      conditions,
      {
        ...update,
        $inc: { ..._.get(update, '$inc', {}), __v: 1 },
      },
      {
        ...options,
        new: true,
      },
    );
  }

  delete({ conditions, options, populates }: DeleteInputInterface): Promise<T> {
    const deletedDoc = this._model.findOneAndDelete(conditions, options);
    return populateHelper<T>(deletedDoc, populates).lean();
  }
}
