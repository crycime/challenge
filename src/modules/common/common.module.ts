import { Global, Module } from '@nestjs/common';
import { LoggerService } from './logger.service';
import { ConfigModule } from '@nestjs/config';
import configuration from '../../config/configuration';

@Global()
@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      load: [configuration],
    }),
  ],
  providers: [
    {
      provide: LoggerService,
      useValue: new LoggerService(),
    },
  ],
  exports: [LoggerService],
})
export class CommonModule {}
