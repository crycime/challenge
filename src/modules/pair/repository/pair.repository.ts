import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { ModelBaseService } from '../../../common/base.service';
import { PairDto } from '../schema/pair.schema';

@Injectable()
export class PairRepository extends ModelBaseService<PairDto> {
  constructor(
    @InjectModel('pair')
    model: Model<PairDto>,
  ) {
    super(model);
  }
}
