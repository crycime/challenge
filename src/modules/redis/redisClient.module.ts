import { Module } from '@nestjs/common';
import { RedisClientService } from './redisClient.service';
import { ConfigService } from '@nestjs/config';
import { RedisOptions } from 'ioredis';
import { RedisModule } from 'nestjs-redis';

@Module({
  imports: [
    RedisModule.forRootAsync({
      useFactory: (configService: ConfigService) => [
        {
          ...configService.get<RedisOptions>('REDIS'),
          retryStrategy: (times: number) => {
            if (times > configService.get('connectionMaxRetry')) {
              console.error({
                location: 'Redis retryStrategy',
                type: 'Redis connection error',
                err: 'Redis connection retry has reach the limit',
              });
              return;
            }
            return Math.min(times * 2000, 3000);
          },
        },
      ],
      inject: [ConfigService],
    }),
  ],
  providers: [RedisClientService],
  exports: [RedisClientService],
})
export class RedisClientModule {}
