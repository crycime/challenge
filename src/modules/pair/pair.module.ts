import { Module } from '@nestjs/common';
import { PairHandlerService } from './services/pair.handler.service';
import { PairController } from './controllers/pair.controller';
import { PairRepository } from './repository/pair.repository';
import { MongooseModule } from '@nestjs/mongoose';
import { PairSchema } from './schema/pair.schema';
import { ScheduleModule } from '@nestjs/schedule';
import { NomicsModule } from '../serviceClient/nomics/nomics.module';
import { PairTasks } from './task/pair.tasks';
import { BullModule } from '@nestjs/bull';
import { PAIR_QUEUE_NAME } from './queue/constants';
import { PairQueue } from './queue/pair.queue';
import { PairProcessor } from './queue/pair.processor';
import { RedisClientModule } from '../redis/redisClient.module';

@Module({
  imports: [
    ScheduleModule.forRoot(),
    MongooseModule.forFeature([{ name: 'pair', schema: PairSchema }]),
    BullModule.registerQueue({
      name: PAIR_QUEUE_NAME,
    }),
    NomicsModule,
    RedisClientModule,
  ],
  controllers: [PairController],
  providers: [
    PairHandlerService,
    PairRepository,
    PairTasks,
    PairQueue,
    PairProcessor,
  ],
})
export class PairModule {}
