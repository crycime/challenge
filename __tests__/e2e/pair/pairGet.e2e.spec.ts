import { HttpStatus, INestApplication } from '@nestjs/common';
import request from 'supertest';
import * as mockingoose from 'mockingoose';
import RedisMock from 'ioredis-mock';
import {
  generateSequentialTests,
  mockGenericApp,
  pairTestModel,
} from '../../helper/e2e.helper';
import { PairModule } from '../../../src/modules/pair/pair.module';
import { PairRepository } from '../../../src/modules/pair/repository/pair.repository';
import { RedisClientService } from '../../../src/modules/redis/redisClient.service';
import { PairHandlerService } from '../../../src/modules/pair/services/pair.handler.service';
import {
  PairSchema,
  pairValidExample,
} from '../../../src/modules/pair/schema/pair.schema';
import { RedisService } from 'nestjs-redis';
import { pairListRedisKey } from '../../../src/utils/redis.util';
import { PairController } from '../../../src/modules/pair/controllers/pair.controller';
import { MongooseModule } from '@nestjs/mongoose';

const path = '/pair';

describe(`GET ${path}`, () => {
  let app: INestApplication;
  let pairHandlerService;
  let pairRepository: PairRepository;
  let redisClientService: RedisClientService;
  let mockRedis: RedisMock;
  let mockFunctionGroup;

  beforeAll(async () => {
    mockRedis = new RedisMock();
    jest
      .spyOn(RedisService.prototype, 'getClient')
      .mockImplementation(() => mockRedis);
    const mockRedisService = {
      getClient: jest.fn(() => mockRedis),
    };
    const appModule = (
      await mockGenericApp({
        module: {
          imports: [
            MongooseModule.forFeature([{ name: 'pair', schema: PairSchema }]),
          ],
          controllers: [PairController],
          providers: [
            PairHandlerService,
            PairRepository,
            { provide: RedisService, useValue: mockRedisService },
            RedisClientService,
          ],
        },
      })
    ).createNestApplication();
    pairHandlerService =
      appModule.get<PairHandlerService>('PairHandlerService');
    pairRepository = appModule.get<PairRepository>('PairRepository');
    redisClientService =
      appModule.get<RedisClientService>('RedisClientService');
    mockFunctionGroup = {
      spyHandlerGetPairList: jest
        .spyOn(pairHandlerService, 'getPairList')
        .mockName('spyHandlerGetPairList'),
      spyRedisGet: jest
        .spyOn(redisClientService, 'get')
        .mockName('spyRedisGet'),
      spyPairRepositoryFind: jest
        .spyOn(pairRepository, 'find')
        .mockName('spyPairRepositoryFind'),
      spyRedisSet: jest
        .spyOn(redisClientService, 'set')
        .mockName('spyRedisSet'),
    };
    app = await appModule.init();
  });

  describe('success cases', () => {
    it('should return pair list response with cache', async () => {
      mockRedis.set(pairListRedisKey(), JSON.stringify([pairValidExample]));
      // mockingoose(pairTestModel).toReturn([pairValidExample], 'find');
      const response = await request(app.getHttpServer()).get(path);
      expect(response.status).toEqual(HttpStatus.OK);
      generateSequentialTests({
        mockFunctionGroup,
        apiResponse: response,
        snapshots: ['spyHandlerGetPairList', 'spyRedisGet'],
        notCalled: ['spyPairRepositoryFind', 'spyRedisSet'],
      });
    });
    it('should return pair list response with no cache', async () => {
      mockingoose(pairTestModel).toReturn([pairValidExample], 'find');
      const response = await request(app.getHttpServer()).get(path);
      expect(response.status).toEqual(HttpStatus.OK);
      generateSequentialTests({
        mockFunctionGroup,
        apiResponse: response,
        snapshots: [
          'spyHandlerGetPairList',
          'spyRedisGet',
          'spyPairRepositoryFind',
          'spyRedisSet',
        ],
      });
    });
  });

  describe('fail cases', () => {
    it('should throw MongoResponse Error if Model find fail', async () => {
      mockingoose(pairTestModel).toReturn(new Error('My Error'), 'find');
      const response = await request(app.getHttpServer()).get(path);
      expect(response.status).toEqual(HttpStatus.BAD_GATEWAY);
      generateSequentialTests({
        mockFunctionGroup,
        apiResponse: response,
        snapshots: [
          'spyHandlerGetPairList',
          'spyRedisGet',
          'spyPairRepositoryFind',
        ],
        notCalled: ['spyRedisSet'],
      });
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    mockingoose.resetAll();
    mockRedis.data.clear();
  });

  afterAll(async () => {
    await app.close();
  });
});
