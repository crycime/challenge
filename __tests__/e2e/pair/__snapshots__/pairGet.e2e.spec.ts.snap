// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /pair fail cases should throw MongoResponse Error if Model find fail 1`] = `
[MockFunction spyHandlerGetPairList] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair fail cases should throw MongoResponse Error if Model find fail 2`] = `
[MockFunction spyRedisGet] {
  "calls": Array [
    Array [
      "pair:list",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair fail cases should throw MongoResponse Error if Model find fail 3`] = `
[MockFunction spyPairRepositoryFind] {
  "calls": Array [
    Array [
      Object {
        "filter": Object {},
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair fail cases should throw MongoResponse Error if Model find fail 4`] = `
Object {
  "code": 502,
  "message": "ModelBaseService - find:My Error",
  "path": "/pair",
  "statusCode": 502,
  "type": "MongoResponseErrorException",
}
`;

exports[`GET /pair success cases should return pair list response with cache 1`] = `
[MockFunction spyHandlerGetPairList] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with cache 2`] = `
[MockFunction spyRedisGet] {
  "calls": Array [
    Array [
      "pair:list",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with cache 3`] = `
Array [
  Object {
    "convert": "USD",
    "currency": "ETH",
    "id": "61cc1f3a428d36921fdcf8d2",
    "name": "Ethereum",
    "price": "3104.46666752",
    "price_change": "34.95363492",
    "volume": "831812243.27",
  },
]
`;

exports[`GET /pair success cases should return pair list response with no cache 1`] = `
[MockFunction spyHandlerGetPairList] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with no cache 2`] = `
[MockFunction spyRedisGet] {
  "calls": Array [
    Array [
      "pair:list",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with no cache 3`] = `
[MockFunction spyPairRepositoryFind] {
  "calls": Array [
    Array [
      Object {
        "filter": Object {},
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with no cache 4`] = `
[MockFunction spyRedisSet] {
  "calls": Array [
    Array [
      Object {
        "key": "pair:list",
        "value": "[{\\"updatedBy\\":\\"system\\",\\"createdBy\\":\\"system\\",\\"_id\\":\\"61cc1f3a428d36921fdcf8d2\\",\\"name\\":\\"Ethereum\\",\\"currency\\":\\"ETH\\",\\"convert\\":\\"USD\\",\\"price\\":\\"3104.46666752\\",\\"volume\\":\\"831812243.27\\",\\"price_change\\":\\"34.95363492\\"}]",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`GET /pair success cases should return pair list response with no cache 5`] = `
Array [
  Object {
    "convert": "USD",
    "createdBy": "system",
    "currency": "ETH",
    "id": "61cc1f3a428d36921fdcf8d2",
    "name": "Ethereum",
    "price": "3104.46666752",
    "price_change": "34.95363492",
    "updatedBy": "system",
    "volume": "831812243.27",
  },
]
`;
